program/src/instruction.rs

```diff
@@ -402,8 +406,74 @@ pub enum StakeInstruction {
     DistributeRewards(u64),
 }
 
+impl StakeInstruction {
+    /// Unpacks a byte buffer into a
+    /// [PaladinRewardsInstruction](enum.PaladinRewardsInstruction.html).
+    pub fn unpack(input: &[u8]) -> Result<Self, ProgramError> {
+        match input.split_first() {
+            // 0 - InitializeConfig: u64 (8) + u16 (2)
+            Some((&0, rest)) if rest.len() == 10 => Ok(StakeInstruction::InitializeConfig {
+                cooldown_time_seconds: i64::from_le_bytes(*array_ref![rest, 0, 8]),
+                max_deactivation_basis_points: u16::from_le_bytes(*array_ref![rest, 8, 2]),
+            }),
+            // 1 - InitializeStake
+            Some((&1, _)) => Ok(StakeInstruction::InitializeStake),
+            // 2- StakeTokens: u64 (8)
+            Some((&2, rest)) if rest.len() == 8 => Ok(StakeInstruction::StakeTokens(
+                u64::from_le_bytes(*array_ref![rest, 0, 8]),
+            )),
+            // 3 - DeactivateStake: u64 (8)
+            Some((&3, rest)) if rest.len() == 8 => Ok(StakeInstruction::DeactivateStake(
+                u64::from_le_bytes(*array_ref![rest, 0, 8]),
+            )),
+            // 4 - InactivateStake
+            Some((&4, _)) => Ok(StakeInstruction::InactivateStake),
+            // 5 - WithdrawInactiveStake: u64 (8)
+            Some((&5, rest)) if rest.len() == 8 => Ok(StakeInstruction::WithdrawInactiveStake(
+                u64::from_le_bytes(*array_ref![rest, 0, 8]),
+            )),
```

Comment:
This is like, super nitty, but it might help clean this function up a bit if you used the brackets and were a tad more verbose on the variants which require further unpacking of `rest`.

This is just an example for `WithdrawInactiveStake`, but it could be useful for the others.

```suggestion
            Some((&5, rest)) if rest.len() == 8 => {
                let amount = u64::from_le_bytes(*array_ref![rest, 0, 8]);
                Ok(StakeInstruction::WithdrawInactiveStake(amount))
            },
```

I only suggest it because the enum has a lot of variants and this function can get a bit hairy - same with its `pack` counterpart. Only if you like it, though. I don't feel too strongly about it.

Reply:
Done!